### PR TITLE
Fix #43906: Support 256th, 512th, and 1024th Notes [WIP]

### PR DESCRIPTION
**Pull request by ajyoon**
- Add glyphs for flags and rests in Mscore and Gootville fonts (@Gootector is it okay to do this as a temporary hack to get support across the different fonts before you do this yourself if you're planning to?) Bravura already includes the required glyphs.
- Partial support for shadow notes. (Bravura and Gootville flags display incorrectly in shadow notes, strangely enough not in Mscore, which is the one I had to touch the most)
- Rework duration validations in various entry points to these durations
- Allow tuplets with these note values
- Allow these durations to be accessed by dots
- Add `importmidi` support for these values
- Update `importmidi` preferences options to allow smallest note value of 1024
- Prevent a fatal assert in `MidiOperations::QuantValue fractionToQuantValue` with a default value
- Misc. code clean-up in the vicinity

Potentially problematic change needing review and feedback: Change `MScore::division` from 480 to ~~512~~ 65536. Current handling of ticks relied on rounding errors when translating `DurationType`'s to ticks, causing bars missing tiny amounts (usually a 1024th value) when pasting sections and similar operations involving the new values. Changing `MScore::division` is also causing most tests to fail currently because of a diff between the embedded division value in the test references and certain tick values diverging. Is this a good way to proceed? If not, are there some better solutions?

TODO's (maybe more appropriate for a separate PR?):
- Add pad icons for new values (currently can be reached by pressing `Q` repeatedly on a note)
- Correct shadow note display for Bravura and Gootville

### CODE REVIEW

**Comment by ajyoon** [comment]
mscore_font
gootville_example

File: fonttools/genft.cpp [hunk 1]
```diff
@@ -81,6 +81,7 @@ QMap<QString, QString> nmap {
       { "rests.5",                  "rest32nd" },
       { "rests.6",                  "rest64th" },
       { "rests.7",                  "rest128th" },
+      { "rests.8",                  "rest256th" },
```
Comment (Jojo-Schmitz) [inline fonttools/genft.cpp]: nothing for 512th and 1024th?

Comment (ajyoon) [inline fonttools/genft.cpp]: Good point - my understanding is this code is intended to create links between Lilypond's Feta/Emmentaler font and the SMuFL compliant Mscore font. Glyphs for `rests.9` and `rests.10` are not defined in Feta. Actually now that I double check [http://lilypond.org/doc/v2.18/Documentation/notation/the-feta-font#rest-glyphs](url) it looks like `rests.8` isn't defined either, so this might be leftover from my testing and I should just cut that line too. Or am I misunderstanding the purpose of this code?

Comment (Jojo-Schmitz) [inline fonttools/genft.cpp]: No idea, I was just wondering...

**Comment by Jojo-Schmitz** [comment]
The mtest failures are all around division and ticks. I wonder whether we really need to adjust all the mtest files or whethr and how to have some backward compatibility to avoid all these failures, which might mean that we can't add these new shorter notes to 2.0 files, only to 3.0 ones?

**Comment by ajyoon** [comment]
I feel like there is a way to do this as a non-breaking change... The whole issue is coming from the fact that ticks have any place at all in the score object model. If all positions were represented as compound fractions this would be a non-issue, and it would also fix a lot of the bugs in tuplets right now as well (and let us fully support arbitrarily nested tuplets too!). But maybe that is outside of the scope of 2.x

**Comment by Jojo-Schmitz** [comment]
I think this change it too big to make it into 2.0.4, so it's be for 3.0 only?

**Comment by lasconic** [comment]
I didn't take a close look yet but two things.
1. If we increase MScore::division, I would increase it enough to accomodate the smaller possible note, including weird tuplets, 512 cannot accommodate a simple 8th note triple (it's not possible to divise it by 3...). Ideally, ticks should be come fractions... but it's out of the score of this PR.
2. The division is saved in the MuseScore file, in the <division> tag. So there should be a way to change the internal MScore::division and use the <division> to do a proper conversion.

**Comment by ajyoon** [comment]
It was actually working before for most common tuples, but certain values (1024th 9-tuplets, etc.) broke things. Increased ``division`` by some arbitrarily high amount and all my testing is working, but this still feels like a band-aid to me.

It seems like there is already some built in way to convert the embedded division value in the scores which might handle the compatibility issue for us. Maybe it's here: https://github.com/musescore/MuseScore/blob/1a8cdf1da4ea63b851001feddf37bb4075c4446c/libmscore/score.h#L740 but I'm having trouble making sense of the purpose of the math so I'm not sure. 

I also added a sanity check to ``ScoreView::cmdCreateTuplet`` to help catch these weird errors sooner.

Seems to me that there won't be a way around making this change without editing mtests...

File: mscore/importmidi/importmidi_operations.cpp [hunk 2]
```diff
@@ -71,6 +71,15 @@ static void setOperationsFromFile(const QString &fileName, Opers &opers)
                                     case 5:
                                           opers.quantValue.setDefaultValue(QuantValue::Q_128, false);
                                           break;
+				    case 6:
+                                          opers.quantValue.setDefaultValue(QuantValue::Q_256, false);
+                                          break;
+				    case 7:
+                                          opers.quantValue.setDefaultValue(QuantValue::Q_512, false);
+                                          break;
+				    case 8:
```
Comment (Jojo-Schmitz) [inline mscore/importmidi/importmidi_operations.cpp]: indentation issue with those case statements

File: libmscore/durationtype.cpp [hunk 3]
```diff
@@ -175,14 +172,10 @@ NoteHead::Type TDuration::headType() const
...
             case DurationType::V_BREVE:
-                  headType = NoteHead::Type::HEAD_BREVIS;
-                  break;
             case DurationType::V_LONG:
                   headType = NoteHead::Type::HEAD_BREVIS;
                   break;
-            default:
-            case DurationType::V_INVALID:
-            case DurationType::V_ZERO:
+	    default:  // V_ZERO or V_INVALID
```
Comment (Jojo-Schmitz) [inline libmscore/durationtype.cpp]: indentation...

File: libmscore/chord.cpp [hunk 4]
```diff
@@ -1240,9 +1240,12 @@ qreal Chord::defaultStemLength()
...
       switch(hookIdx) {
-            case 3: normalStemLen += small() ? .5  : 0.75; break; //32nd notes
-            case 4: normalStemLen += small() ? 1.0 : 1.5;  break; //64th notes
-            case 5: normalStemLen += small() ? 1.5 : 2.25; break; //128th notes
+            case 3: normalStemLen += small() ? .5  : 0.75; break; // 32nd notes
+            case 4: normalStemLen += small() ? 1.0 : 1.5;  break; // 64th notes
+            case 5: normalStemLen += small() ? 1.5 : 2.25; break; // 128th notes
+            case 6: normalStemLen += small() ? 2   : 3;    break; // 256th notes
+            case 7: normalStemLen += small() ? 2.5 : 3.75; break; // 512th notes
+            case 8: normalStemLen += small() ? 3   : 4.5;  break; // 1024th notes
```
Comment (Jojo-Schmitz) [inline libmscore/chord.cpp]: No default case? No 8th or 16th?

File: mscore/scoreview.cpp [hunk 5]
```diff
@@ -4882,6 +4881,10 @@ void ScoreView::cmdCreateTuplet( ChordRest* cr, Tuplet* tuplet)
                   sm->postEvent(new CommandEvent("note-input"));
             _score->inputState().setDuration(tuplet->baseLen());
             }
+#ifndef NDEBUG
```
Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: why only in DEBUG builds?

File: libmscore/chord.cpp (hunk 4 above)
Comment (ajyoon) [inline libmscore/chord.cpp]: default case is line 1241 - 8ths and 16ths have the same length. (I didn't touch this beside adding the new values)

Comment (Jojo-Schmitz) [inline libmscore/chord.cpp]: OK, I was just wondering

File: mscore/scoreview.cpp (hunk 5 above)
Comment (ajyoon) [inline mscore/scoreview.cpp]: Copied from https://github.com/musescore/MuseScore/blob/master/libmscore/layout.cpp#L3580 - should we prefer to do it in all builds in every case?

Comment (Jojo-Schmitz) [inline mscore/scoreview.cpp]: not sure, @lasconic?

**Comment by Gootector** [comment]
1024th, 512th and 256th notes will be added by me in the next version of the Gootville font - 1.4. But... Who are using these notes?

**Comment by ajyoon** [comment]
@Gootector Is it ok to use my slightly hacky glyphs in the meantime before you finish 1.4 just so we can have everything supported in the meantime? 

These values appear in Beethoven even - https://en.wikipedia.org/wiki/Two_hundred_fifty-sixth_note - as well as _lots_ of contemporary music. The aim is to make this environment more permissive for composers so we don't decide for them what they should or shouldn't notate (to be honest, personally I would never write these values, but many of my peers do). It's an easy enough feature to implement with all the scaffolding already in place.

**Comment by Gootector** [comment]
@ajyoon I'll add 256th-1024th noteflags and rests to the font files - as a new glyphs. No problem for me. If it's necessary...
OK. Let's assume that 256th note is necessary. In MuseScore 4.0: 2048th and 4096th? I've a problem with reading 256th note in scores... 1024th - this note is unreadable for me and should exist in theory only. It's my opinion, of course.

**Comment by Jojo-Schmitz** [comment]
Bravura supports up to 1024th, so should any other SMuFL font. More shouldn't be needed

**Comment by lasconic** [comment]
Increasing the division to 65536 will limit the length of score to MAX_INTEGER/65536=  32768 quarters notes, = 8192 measures in 4/4. (assuming MAX_INTEGER = 2^31) 
@wschweer thinks it's a bit low. Can we find a middle ground?

**Comment by Jojo-Schmitz** [comment]
how about 26880, it divides by 256, 3, 5 and 7, so we should be fine up until  10-lets (only x-lets with x being a prime number >= 11 won't work) of 1024th and allow more than double the lenght of a score than with 65536 (which still won't allow triplets, 5-lets or 7-lets of 1024th)?
Or is my math and/or logic flawed?

**Comment by mgavioli** [comment]
I was about to comment along @Jojo-Schmitz line: 65536 is only divisible by 2 (several times) and does not accommodate any tuplet at all.

I was going to suggest 9 * 7 * 5 = 315 * a convenient amount of powers of 2; for instance, 315 * 32 = 10080 or 315 * 64 = 20160. 9 is required to allow 9-plets.

**Comment by Jojo-Schmitz** [comment]
Ah, yes, I forgot 9-lets.

**Comment by Gootector** [comment]
Note values in commercial scorewriters:
Finale - max 128th
Sibelius - max 512th

**Comment by lasconic** [comment]
Extreme of conventional notation http://homes.soic.indiana.edu/donbyrd/CMNExtremesBody.htm#duration

For 1024th and 3,5,6,9 tuplets, we would need 256*315=80640...

**Comment by mgavioli** [comment]
Indeed! But it is so important to support tuplets of 1024th?

315*64 = 20160 would support tuplets of 256th (1/4 / 1/256 = 64), which is already quite esoteric...

**Comment by lasconic** [comment]
Mmm, we still need to be able to represent 1024th, so the number needs to be divisible by 256 right?

**Comment by mgavioli** [comment]
Ops, yes!

**Comment by ajyoon** [comment]
To be honest I don't understand the mechanics behind how tuplets interract with ticks, since my tests with these values (even triplets with the previously tested `division` of `512`, which does not divide evenly) were working. It seems the tick `division` doesn't need to be perfectly divisible by the note fraction to work. Some kind of sneaky rounding voodoo is suspected.

**Comment by MarcSabatella** [comment]
There is indeed "voodoo" scattered throughout the code - various fudge factors worked in to tick calculations.  But they only work up to a point, so it would be nice to not have to rely on these so much.

**Comment by ajyoon** [comment]
Thanks everyone for the feedback! I've reduced the `division` value to 4096 per @lasconic and my testing with creating tuplets safely and all is all still working. Of course mtests are still failing. Does anyone have some idea how to update mtests safely here? It seems we'd just need to update the saved `division` tag and update `tick` values.

**Comment by ajyoon** [comment]
Apologies for the delay on this, I spent the night taking another crack at this but after rebase the tests are failing even worse than before. I suspect b85cad9d8331b9cc4310a156bb70a1ad61cae617 may be involved, but I'm not entirely sure. I tried rebuilding the tests using this branch with a script that uses the build to open and save every `.mscx` file in `mtest` except for those in `compat` which aren't `-ref` files, which brought the failure rate to more like 50% but even still it seems not to be a very safe way to go about it. Is this the right idea for how tests should be rebuilt? If not, what might be a good way to go about this?

**Comment by ajyoon** [comment]
If this doesn't seem worth pursuing for the compatibility risk for such a trivial feature, please feel free to close and I'll work on something more useful

File: libmscore/sym.cpp [hunk 6]
```diff
@@ -5321,6 +5321,7 @@ QVector<oldName> oldNames = {
       {"32' rest",                              SymId::rest32nd },            // rests.5
       {"64' rest",                              SymId::rest64th },            // rests.6
       {"128' rest",                             SymId::rest128th },           // rests.7
+      {"256' rest",                             SymId::rest256th },           // rests.8
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp]: Where's 512th and 1024th rest?

File: fonttools/genft.cpp [hunk 7]
```diff
@@ -81,6 +81,9 @@ QMap<QString, QString> nmap {
       { "rests.5",                  "rest32nd" },
       { "rests.6",                  "rest64th" },
       { "rests.7",                  "rest128th" },
+//    { "rests.8",                  "rest256th" },  // Glyphs currently don't exist in Emmentaler
```
Comment (Jojo-Schmitz) [inline fonttools/genft.cpp]: so these need to get added

File: fonts/mscore/glyphnames.json [hunk 8]
```diff
@@ -817,7 +841,19 @@
...
+        "description": "1024th rest"
+    },
+    "rest512th": {
+        "codepoint": "U+E4EC", 
+        "description": "512th rest"
+    },
+    "rest256th": {
+        "codepoint": "U+E4EB", 
+        "description": "256th (demisemihemidemisemiquaver) rest"
+    },
```
Comment (Jojo-Schmitz) [inline fonts/mscore/glyphnames.json]: We need to be aware of PR #1740, if and when that finally gets finished

File: libmscore/sym.cpp (hunk 6 above)
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp]: Where's 512th and 1024th rest? Probably not needed as they can't happen in 1.x score, so this change doesn't seem to be needed at all.
 As we recently dropped 1.x score support for 3.0, we may even get rid of the entire oldName stuff. Unless there are plans to have this PR getting integrated into the 2.0.4 branch too.

File: fonttools/genft.cpp (hunk 7 above)
Comment (ajyoon) [inline fonttools/genft.cpp]: I'm wondering if `genft` is actually doing anything at all at this point. It seems like it was used previously when converting the original Emmantaler to the MuseScore font, but since the two have forked I don't really understand what this code does anymore. Do we still need it at all?

File: libmscore/sym.cpp (hunk 6 above)
Comment (ajyoon) [inline libmscore/sym.cpp]: I'll put it in and also do the sync with `fonts/mscore/glyphnames.json` (totally missed this one!) and if we decide not to put it in with 2.0.4 then it can be cut. But I think if we do end up using this PR it would be best to aim for 2.0.4/5 since the difference in tick division will make it a huge pain to keep tests synchronized between branches otherwise.

Comment (Jojo-Schmitz) [inline libmscore/sym.cpp]: Could be taken out in a separate commit, one that just doesn't get cherry-picked for 2.0.4

Comment (Jojo-Schmitz) [inline libmscore/sym.cpp]: I do believe though that this PR is a too big and risky change for any branch of 2.0

Comment (ajyoon) [inline libmscore/sym.cpp]: I agree it is very risky and the tests are structured that it's hard to know if it's really safe. If we're targeting for 3.0 maybe it's better to put this on the backburner for a while and, as another project, move the entire object model off of ticks entirely. I think the whole headache here is really indicative that it would be much better to just represent durations and locations purely in terms of rational fractions. It would make future changes like this much, much easier, and it would also allow us to easily support arbitrarily nested tuples with far less need for sanity checks and relying on rounding errors(!!!).

**Comment by Jojo-Schmitz** [comment]
Rebase needed. And thanks to the new fallback, the added glyphs for MScore and Gootville are not vital anymore (but would still be appreciated)

**Comment by ajyoon** [comment]
Rebased with new smufl mapping accounted for

**Comment by Jojo-Schmitz** [comment]
See also #3102 and #3107

**Comment by smoge** [comment]
The software should allow the publisher or composer to make those decisions, the developers should provide the framework for them to decide this... 

In the repertoire, even more radical notation have existed since the 17th century.

> The shortest note value to have ever been used in a published work is the  1⁄1024 note (notated incorrectly as a  1⁄2048) in Anthony Philip Heinrich's Toccata Grande Cromatica from The Sylviad, Set 2, written around 1825;  1⁄256 notes occur frequently in this piece, and some  1⁄512 notes also appear.[2] For comparison, the shortest notated duration supported by Finale is a  1⁄4096 note,[7] while LilyPond can write notes with up to 27 beams, as short as a  1⁄1073741824 (= 2−30) note

[https://en.wikipedia.org/wiki/Two_hundred_fifty-sixth_note)](https://en.wikipedia.org/wiki/Two_hundred_fifty-sixth_note)

**Comment by Jojo-Schmitz** [comment]
Rebase needed

**Comment by Jojo-Schmitz** [comment]
I wonder whether in the light of #3812 the setting tor `MScore::division` becomes lees of an issue

**Comment by anatoly-os** [comment]
The shortest notes are fully supported now thanks to the switching to Fractions.

**Comment by ajyoon** [comment]
Awesome, thanks!

**Comment by Jojo-Schmitz** [comment]
Still we don't have notes that short, still 128th is the shorted note we can enter, except via entering a 128th and than using Q to half its duration
